pkg/deploy/deployer.go

```diff
@@ -115,25 +126,78 @@ func (d *githubDeployer) Deploy(serviceNames []string, environmentName, commit,
+		}
+		if frozen {
+			frozenServices = append(frozenServices, service.Name)
+		}
 	}
 
-	servicesString := strings.Join(serviceNames, ",")
+	if len(frozenServices) > 0 {
+		return nil, "", api.NewValidationErr(fmt.Sprintf("cannot deploy: services are frozen: %s", strings.Join(frozenServices, ", ")))
+	}
 
 	logWithCtx.Infof("Starting deployment")
-	branch := fmt.Sprintf("deploy-%s-%s", servicesString, environmentName)
 	prTitle := fmt.Sprintf("Deploy %s to %s with version %s triggered by %s (%s)", servicesString, environmentName, commit[:7], userFullname, userEmail)
 
-	baseFolder, err := os.MkdirTemp(d.config.Github.CloneTmpDir, branch+"-*")
+	for service, environment := range serviceToEnvironment {
+		files, err := d.renderTemplates(baseFolder, environment.TemplatePath, environment.GeneratedPath, service.Name, environmentName, commit)
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to render templates for service %s, error: %w", service.Name, err)
+		}
+
+		tree, err := d.githubClient.CreateTree(ctx, ref, baseFolder, files)
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to create diff tree for service %s, error: %w", service.Name, err)
+		}
+
+		commitMsg := fmt.Sprintf("Deploy %s to %s with version %s triggered by %s (%s)", service.Name, environmentName, commit[:7], userFullname, userEmail)
+		if err = d.githubClient.PushCommit(ctx, ref, tree, userFullname, userEmail, commitMsg); err != nil {
+			return nil, "", fmt.Errorf("failed to create commit for service %s, error: %w", service.Name, err)
+		}
+	}
+
+	prDescription := fmt.Sprintf("Service Names: %s\nEnvironment: %s\nCommit: [%s](%s)\nRequested by: %s (%s)",
+		servicesString, environmentName, commit[:7], commitUrl, userFullname, userEmail)
+	pr, diff, err := d.githubClient.CreatePR(ctx, prTitle, prDescription, deploymentBranch, branch)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to create pull request, error: %w", err)
+	}
+
+	logWithCtx.Infof("Created pull request for deployment")
+
+	return pr, diff, nil
+}
+
+func (d *githubDeployer) Freeze(serviceNames []string, environment, userFullname, userEmail string, action FreezeAction) (*github.PullRequest, string, error) {
+	ctx := context.Background()
+	logWithCtx := log.WithFields(log.Fields{
+		"environment":  environment,
+		"serviceNames": serviceNames,
+		"action":       action,
+	})
+
+	serviceToEnvironment, deploymentBranch, err := d.resolveServicesAndEnvironment(serviceNames, environment)
 	if err != nil {
-		return nil, "", fmt.Errorf("failed to create temp directory, error: %w", err)
+		return nil, "", err
 	}
 
-	ref, err := d.githubClient.Clone(ctx, deploymentBranch, branch, baseFolder)
+	servicesString := strings.Join(serviceNames, ",")
+
+	logWithCtx.Infof("Starting %s operation", action)
+	branch := fmt.Sprintf("%s-%s-%s", action, servicesString, environment)
+	prTitle := fmt.Sprintf("%s %s to %s triggered by %s (%s)", action, servicesString, environment, userFullname, userEmail)
```

Comment:
lets move pr title and branch names creation to methods and reuse

Reply:
the clone and create PR is already shared functions